temporian/implementation/numpy_cc/operators/window.cc

```diff
@@ -10,436 +10,566 @@
+py::array_t<OUTPUT> accumulate(const ArrayD &evset_timestamps,
+                               const py::array_t<INPUT> &evset_values,
+                               const ArrayD &sampling_timestamps,
+                               const double window_length) {
+  // Input size
+  const size_t n_event = evset_timestamps.shape(0);
+  const size_t n_sampling = sampling_timestamps.shape(0);
+
+  // Allocate output array
+  auto output = py::array_t<OUTPUT>(n_sampling);
+
+  auto v_output = output.template mutable_unchecked<1>();
+  auto v_timestamps = evset_timestamps.unchecked<1>();
+  auto v_values = evset_values.template unchecked<1>();
+  auto v_sampling = sampling_timestamps.unchecked<1>();
+
+  TAccumulator accumulator;
 
-      v_output[sampling_idx] = accumulator.Result();
+  size_t begin_idx = 0;
+  size_t end_idx = 0;
+
+  for (size_t sampling_idx = 0; sampling_idx < n_sampling; sampling_idx++) {
+    const auto right_limit = v_sampling[sampling_idx];
+
+    while (end_idx < n_event && v_timestamps[end_idx] <= right_limit) {
+      accumulator.Add(v_values[end_idx]);
+      end_idx++;
     }
 
-    return output;
+    while (begin_idx < n_event &&
+           // Compare both sides around ~0 to get maximum float resolution
+           v_sampling[sampling_idx] - v_timestamps[begin_idx] >=
+               window_length) {
+      accumulator.Remove(v_values[begin_idx]);
+      begin_idx++;
+    }
+
+    v_output[sampling_idx] = accumulator.Result();
   }
 
-  // Note: We only use inheritence to compile check the code.
-  template <typename INPUT, typename OUTPUT>
-  struct Accumulator
-  {
-    virtual ~Accumulator() = default;
-    virtual void Add(INPUT value) = 0;
-    virtual void Remove(INPUT value) = 0;
-    virtual OUTPUT Result() = 0;
-  };
-
-  template <typename INPUT, typename OUTPUT>
-  struct SimpleMovingAverageAccumulator : Accumulator<INPUT, OUTPUT>
-  {
-    void Add(INPUT value) override
-    {
-      if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN)
-      {
-        if (std::isnan(value))
-        {
-          return;
+  return output;
+}
+
+// No external sampling, variable window length
+template <typename INPUT, typename OUTPUT, typename TAccumulator>
+py::array_t<OUTPUT> accumulate(const ArrayD &evset_timestamps,
+                               const py::array_t<INPUT> &evset_values,
+                               const ArrayD &window_length) {
+  // Input size
+  const size_t n_event = evset_timestamps.shape(0);
+
+  // Allocate output array
+  auto output = py::array_t<OUTPUT>(n_event);
+
+  auto v_output = output.template mutable_unchecked<1>();
+  auto v_timestamps = evset_timestamps.unchecked<1>();
+  auto v_values = evset_values.template unchecked<1>();
+  auto v_window_length = window_length.unchecked<1>();
+
+  TAccumulator accumulator;
+
+  // Index of the first value in the window.
+  size_t begin_idx = 0;
+  // Index of the first value outside the window.
+  size_t end_idx = 0;
+
+  double curr_ts;

```

Comment:
Nit: here, there might be only limited benefit to declaring the variable outside of the while.

Reply:
Mind expanding on that? Although a quick search reveals a possible answer https://stackoverflow.com/questions/7959573/declaring-variables-inside-loops-good-practice-or-bad-practice (TLDR scoping the variable correctly lets compiler do better optimization + the variable isn't re-allocated on each loop, only once). Changed all declarations to `const auto` inside the loop

Reply:
Nice :).
Note however that it depends on the object (and the cost of construction / destruction). The person asks about string and the first answer is about ints.
For example, with simple primitives (e.g., double), it doesn't matter. All three functions compile to the same code: https://godbolt.org/z/qboWse5v4
However, if you have an expensive object that can be reused, it should be put outside of the loop (see second answer in your link).